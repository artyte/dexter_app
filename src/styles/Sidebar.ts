import styled from 'styled-components'

export const Button = styled.button`
  background-color: Transparent;
  background-repeat: no-repeat;
  border: none;
  cursor: pointer;
  overflow: hidden;
  outline: none;
`

export const TopIcons = styled.li`
  list-style-type: none;
  margin-top: 15px;
`

export const BottomIcons = styled.li`
  list-style-type: none;
  text-align: center;
  margin-bottom: 15px;
  position: absolute;
  bottom: 0;
`

export const Nav = styled.nav`
  background-color: #e74c3c;
  padding-left: 15px;
  padding-right: 15px;
  font-size: 100%;
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  filter: ${(props: any) => (props.blur ? 'blur(5px)' : 'blur(0px)')};
`
