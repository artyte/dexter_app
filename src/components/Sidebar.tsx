import React, { memo, useContext, useState, useCallback } from 'react'
import profile from '../assets/user.svg'
import pc from '../assets/pc.svg'
import team from '../assets/team.svg'
import port from '../assets/port.svg'
import login from '../assets/login.svg'
import logout from '../assets/logout.svg'
import UserContext from '../context/User'
import { Nav, Button, TopIcons, BottomIcons } from '../styles/Sidebar'

/**
 * @param blur: The state of whether to blur this page so that window popups can happen.
 * @param toggleBox: Toggles between team page and teams (otherwise known as PC) page.
 * @param toggleProfile: Toggles the popup for the user profile's page.
 * @param togglePort: Toggles the popup for the import/export page.
 * @param toggleLogin: Toggles the popup for the login page.
 * @param testNavBar: Test id for testing navbar
 * @param testSwitchBox: Test id for testing team/pc button
 */
type props = {
  blur: boolean
  toggleBox: () => void
  toggleProfile: () => void
  togglePort: () => void
  toggleLogin: () => void
  testNavBar?: string
  testSwitchBox?: string
  testToggleProfile?: string
  testTogglePort?: string
  testToggleLogin?: string
}
/**
 * This is the sidebar of the web app, containing links to popups and toggling between 2 pages.
 *
 * @type props: Refer to type props
 */
const Sidebar = ({
  blur,
  toggleBox,
  toggleProfile,
  togglePort,
  toggleLogin,
  testNavBar,
  testSwitchBox,
  testToggleProfile,
  testTogglePort,
  testToggleLogin,
}: props) => {
  const { token } = useContext(UserContext)

  const [boxAlt, setBoxAlt] = useState('team')
  const [boxSrc, setBoxSrc] = useState(team)

  /** Returns the clickable profile icon. */
  const makeProfile = () => (
    <TopIcons>
      <Button onClick={toggleProfile} data-testid={testToggleProfile ?? ''}>
        <img src={profile} alt='profile' />
      </Button>
    </TopIcons>
  )

  /** Toggles between the icon of team or teams (otherwise known as pc) page. */
  const changeTeam = useCallback(() => {
    if (boxAlt === 'team') {
      setBoxAlt('pc')
      setBoxSrc(pc)
    } else if (boxAlt === 'pc') {
      setBoxAlt('team')
      setBoxSrc(team)
    }
    toggleBox()
  }, [boxAlt, toggleBox])

  /** Returns the clickable team/pc icon. */
  const makeBox = () => (
    <TopIcons>
      <Button onClick={changeTeam} data-testid={testSwitchBox ?? ''}>
        <img src={boxSrc} alt={boxAlt} />
      </Button>
    </TopIcons>
  )

  /** Returns the clickable import/export icon. */
  const makePort = () => (
    <TopIcons>
      <Button onClick={togglePort} data-testid={testTogglePort ?? ''}>
        <img src={port} alt='port' />
      </Button>
    </TopIcons>
  )

  /** Returns the clickable login or logout icon. */
  const makeAccess = () => (
    <BottomIcons>
      <Button onClick={toggleLogin} data-testid={testToggleLogin ?? ''}>
        <img src={token ? logout : login} alt={token ? 'logout' : 'login'} />
      </Button>
    </BottomIcons>
  )

  return (
    <Nav blur={blur} data-testid={testNavBar ?? ''}>
      {makeProfile()}
      {makeBox()}
      {makePort()}
      {makeAccess()}
    </Nav>
  )
}

export default memo(Sidebar)
