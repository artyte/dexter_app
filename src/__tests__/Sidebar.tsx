import React, { useState } from 'react'
import { BrowserRouter as Router } from 'react-router-dom'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { toHaveStyle } from '@testing-library/jest-dom/matchers'
import Sidebar from '../components/Sidebar'

expect.extend({ toHaveStyle })

afterEach(cleanup)

const empty = jest.fn()

describe('Render navbar', () => {
  describe('with blur', () => {
    it('in team page', () => {
      const testNavId = 'navBar'
      const testToggleProfileId = 'toggleProfile'
      const SidebarWrapper = () => {
        const [blur, setBlur] = useState(false)

        return (
          <Router>
            <Sidebar
              togglePort={empty}
              toggleLogin={empty}
              toggleProfile={jest.fn(() => setBlur(!blur))}
              blur={blur}
              testNavBar={testNavId}
              testToggleProfile={testToggleProfileId}
            />
          </Router>
        )
      }
      const { getByTestId, asFragment } = render(
        <SidebarWrapper />
      )

      expect(getByTestId(testNavId)).toHaveStyle('filter: blur(0px);')

      expect(asFragment()).toMatchSnapshot()
      
      fireEvent.click(getByTestId(testToggleProfileId))
      expect(asFragment()).toMatchSnapshot()
      expect(getByTestId(testNavId)).toHaveStyle('filter: blur(5px);')
    })

    it('in pc page', () => {
      const testNavId = 'navBar'
      const testToggleProfileId = 'toggleProfile'
      const testSwitchBoxId = 'switchBox'
      const SidebarWrapper = () => {
        const [blur, setBlur] = useState(false)

        return (
          <Router>
            <Sidebar
              togglePort={empty}
              toggleLogin={empty}
              toggleProfile={jest.fn(() => setBlur(!blur))}
              blur={blur}
              testNavBar={testNavId}
              testToggleProfile={testToggleProfileId}
              testSwitchBox={testSwitchBoxId}
            />
          </Router>
        )
      }
      const { getByTestId, asFragment } = render(<SidebarWrapper />)

      fireEvent.click(getByTestId(testSwitchBoxId))
      expect(getByTestId(testNavId)).toHaveStyle('filter: blur(0px);')

      expect(asFragment()).toMatchSnapshot()

      fireEvent.click(getByTestId(testToggleProfileId))
      expect(asFragment()).toMatchSnapshot()
      expect(getByTestId(testNavId)).toHaveStyle('filter: blur(5px);')
    })
  })

  it('without blur', () => {
    const testNavId = 'navBar'
    const testSwitchBoxId = 'switchBox'
    const { getByTestId, asFragment } = render(
      <Router>
        <Sidebar
          toggleProfile={empty}
          togglePort={empty}
          toggleLogin={empty}
          blur={false}
          testNavBar={testNavId}
          testSwitchBox={testSwitchBoxId}
        />
      </Router>
    )
    expect(asFragment()).toMatchSnapshot()
    expect(getByTestId(testNavId)).toHaveStyle('filter: blur(0px);')

    fireEvent.click(getByTestId(testSwitchBoxId))
    expect(asFragment()).toMatchSnapshot()
    expect(getByTestId(testNavId)).toHaveStyle('filter: blur(0px);')
  })
})
