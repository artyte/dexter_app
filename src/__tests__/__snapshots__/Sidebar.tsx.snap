// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render navbar with blur in pc page 1`] = `
<DocumentFragment>
  <nav
    class="sc-AxhCb umvJw"
    data-testid="navBar"
  >
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid="toggleProfile"
      >
        <img
          alt="profile"
          src="user.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid="switchBox"
      >
        <img
          alt="pc"
          src="pc.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="port"
          src="port.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxiKw dfPWIn"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="login"
          src="login.svg"
        />
      </button>
    </li>
  </nav>
</DocumentFragment>
`;

exports[`Render navbar with blur in pc page 2`] = `
<DocumentFragment>
  <nav
    class="sc-AxhCb gTisAF"
    data-testid="navBar"
  >
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid="toggleProfile"
      >
        <img
          alt="profile"
          src="user.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid="switchBox"
      >
        <img
          alt="pc"
          src="pc.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="port"
          src="port.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxiKw dfPWIn"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="login"
          src="login.svg"
        />
      </button>
    </li>
  </nav>
</DocumentFragment>
`;

exports[`Render navbar with blur in team page 1`] = `
<DocumentFragment>
  <nav
    class="sc-AxhCb umvJw"
    data-testid="navBar"
  >
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid="toggleProfile"
      >
        <img
          alt="profile"
          src="user.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="team"
          src="team.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="port"
          src="port.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxiKw dfPWIn"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="login"
          src="login.svg"
        />
      </button>
    </li>
  </nav>
</DocumentFragment>
`;

exports[`Render navbar with blur in team page 2`] = `
<DocumentFragment>
  <nav
    class="sc-AxhCb gTisAF"
    data-testid="navBar"
  >
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid="toggleProfile"
      >
        <img
          alt="profile"
          src="user.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="team"
          src="team.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="port"
          src="port.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxiKw dfPWIn"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="login"
          src="login.svg"
        />
      </button>
    </li>
  </nav>
</DocumentFragment>
`;

exports[`Render navbar without blur 1`] = `
<DocumentFragment>
  <nav
    class="sc-AxhCb umvJw"
    data-testid="navBar"
  >
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="profile"
          src="user.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid="switchBox"
      >
        <img
          alt="team"
          src="team.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="port"
          src="port.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxiKw dfPWIn"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="login"
          src="login.svg"
        />
      </button>
    </li>
  </nav>
</DocumentFragment>
`;

exports[`Render navbar without blur 2`] = `
<DocumentFragment>
  <nav
    class="sc-AxhCb umvJw"
    data-testid="navBar"
  >
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="profile"
          src="user.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid="switchBox"
      >
        <img
          alt="pc"
          src="pc.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxirZ fJJekH"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="port"
          src="port.svg"
        />
      </button>
    </li>
    <li
      class="sc-AxiKw dfPWIn"
    >
      <button
        class="sc-AxjAm dkMHIj"
        data-testid=""
      >
        <img
          alt="login"
          src="login.svg"
        />
      </button>
    </li>
  </nav>
</DocumentFragment>
`;
